src/sentry/rules/actions/utils.py

```diff
@@ -0,0 +1,96 @@
+from collections import defaultdict
+from typing import Any
+
+from sentry.api.serializers.models.rule import generate_rule_label
+from sentry.models.environment import Environment
+from sentry.models.rule import Rule
+
+
+def get_changed_data(
+    rule: Rule, rule_data: dict[str, list[Any]], rule_data_before: dict[str, list[Any]]
+) -> dict[str, list[Any]]:
+    """
+    Generate a list per type of issue alert rule data of what changes occurred on edit.
+    """
+    changed_data = defaultdict(list)
+
+    for condition in rule_data.get("conditions", []):
+        if condition not in rule_data_before.get("conditions", []):
+            label = generate_rule_label(rule_data.get("project"), rule, condition)
+            changed_data[condition["id"]].append(f"Added condition '{label}'")
+
+    for condition in rule_data_before.get("conditions", []):
+        if condition not in rule_data.get("conditions", []):
+            label = generate_rule_label(rule.project, rule, condition)
+            changed_data[condition["id"]].append(f"Removed condition '{label}'")
+
+    for action in rule_data.get("actions", []):
+        if action not in rule_data_before.get("actions", []):
+            label = generate_rule_label(rule.project, rule, action)
+            changed_data[condition["id"]].append(f"Added action '{label}'")
+
+    for action in rule_data_before.get("actions", []):
+        if action not in rule_data.get("actions", []):
+            label = generate_rule_label(rule.project, rule, action)
+            changed_data[condition["id"]].append(f"Removed action '{label}'")
+
+    if rule_data.get("frequency") != rule_data_before.get("frequency"):
+        old_frequency = rule_data_before.get("frequency")
+        new_frequency = rule_data.get("frequency")
+        changed_data["changed_frequency"].append(
+            f"Changed frequency from *{old_frequency}* to *{new_frequency}*"
+        )
```

Comment:
It could also be worth considering this logic as a helper, as we repeat it 4 times